### PR TITLE
Add ScopedKind to IParameterSymbol and ILocalSymbol

### PR DESCRIPTION
**Pull request by cston**
See https://github.com/dotnet/roslyn/issues/61647.

### CODE REVIEW

File: src/Compilers/Core/Portable/Symbols/ScopedKind.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Enumeration for kinds of scoped modifiers.
+    /// </summary>
+    public enum ScopedKind
```
Comment (cston) [inline src/Compilers/Core/Portable/Symbols/ScopedKind.cs:10]: >ScopedKind [](http://example.com/codeflow?start=16&length=10)

Should this be `: byte`?

File: src/Compilers/CSharp/Portable/Symbols/DeclarationScope.cs [hunk 2]
```diff
@@ -2,16 +2,28 @@
...
 
+using Roslyn.Utilities;
+
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
-    // https://github.com/dotnet/roslyn/issues/61647: Internally, scope is represented with this enum,
-    // but the public API uses a pair of IsRefScoped and IsValueScoped bools (see ILocalSymbol,
-    // IParameterSymbol, and ScopedRefAttribute). We should have a common representation.
-    // And we should use common terms for the attribute and enum names.
     internal enum DeclarationScope : byte
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/DeclarationScope.cs:9]: >DeclarationScope [](http://example.com/codeflow?start=18&length=16)

Replace with `ScopedKind`?

File: src/Compilers/Core/Portable/Symbols/IParameterSymbol.cs [hunk 3]
```diff
@@ -22,6 +22,11 @@ public interface IParameterSymbol : ISymbol
         /// </summary>
         RefKind RefKind { get; }
 
+        /// <summary>
+        /// Returns the scoped kind of the local.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/IParameterSymbol.cs:26]: >local [](http://example.com/codeflow?start=43&length=5)

parameter?

File: src/Compilers/Core/Portable/Symbols/ScopedKind.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    /// </summary>
+    public enum ScopedKind : byte
+    {
+        /// <summary>
+        /// Not scoped.
+        /// </summary>
+        None = 0,
+
+        /// <summary>
+        /// A ref scoped to the current method or block.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/ScopedKind.cs:18]: >current [](http://example.com/codeflow?start=32&length=7)

enclosing?

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs [hunk 5]
```diff
@@ -1129,9 +1115,8 @@ private void AddParameterRefKindIfNeeded(IParameterSymbol symbol)
         {
             if (format.ParameterOptions.IncludesOption(SymbolDisplayParameterOptions.IncludeParamsRefOut))
             {
-                // https://github.com/dotnet/roslyn/issues/61647: Use public API.
-                if ((symbol as Symbols.PublicModel.ParameterSymbol)?.GetSymbol<ParameterSymbol>() is { } parameter &&
-                    parameter.EffectiveScope == DeclarationScope.RefScoped &&
+                if (symbol.ScopedKind == ScopedKind.ScopedRef &&
+                    (symbol as Symbols.PublicModel.ParameterSymbol)?.GetSymbol<ParameterSymbol>() is { } parameter &&
                     !ParameterHelpers.IsRefScopedByDefault(parameter) &&
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:1120]: Are we missing ```IsRefScopedByDefault``` as a public API? Is there an alternative way to detect this in a language independent way?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:1120]: If so, perhaps we should use this condition only for C# symbols and not dismiss other symbols?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:1120]: `IsRefScopedByDefault()` relies on `ModuleSymbol.UsesUpdatedEscapeRules`, which is not tracked in VB, nor is `scoped`. Changed to use for C# symbols only.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4)

File: src/Compilers/CSharp/Portable/Symbols/DeclarationScope.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/DeclarationScope.cs:9]: Seems like it would be a reasonable change.

**Review by 333fred — APPROVED** [review]
LGTM (commit 4). It does seem like it would be reasonable to remove `DeclarationScope` and just use `ScopedKind` everywhere, imo.

File: src/Compilers/CSharp/Portable/Symbols/DeclarationScope.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/DeclarationScope.cs:9]: https://github.com/dotnet/roslyn/issues/64248

**Pull request merged**